File: backend-node/controllers/userController.js
// controllers/usersController.js
const bcrypt = require('bcrypt');
const Users = require('../models/usersModel');

let loggedInUsers = {}; // In-memory store for logged-in users

exports.register = (req, res) => {
    const { username, email, password } = req.body;
    const salt = bcrypt.genSaltSync(10);
    const hashedPassword = bcrypt.hashSync(password, salt);

    const newUser = {
        username,
        email,
        pass: hashedPassword,
        salt: salt,
    };

    Users.create(newUser, (err, data) => {
        if (err) return res.status(500).send(err);
        const userId = data.id;
        res.status(201).json({ user_id: userId, message: 'User registered successfully' });
    });
};

exports.login = (req, res) => {
    const { username, password } = req.body;

    const findUserCallback = (err, users) => {
        if (err || users.length === 0) return res.status(400).send('User not found');

        const user = users[0];
        console.log(user);
        const hashedPassword = bcrypt.hashSync(password, user.salt);
    
        console.log(hashedPassword)
        if (hashedPassword !== user.pass) {
            return res.status(400).send('Invalid password');
        } else{
            // Store the logged-in user ID
            loggedInUsers[user.id] = true;
            res.status(200).json({ message: 'Login successful', userId: user.id });
        } 
    };

    if (username) {
        Users.findByUsername(username, findUserCallback);
    } else {
        res.status(400).send('Username is required');
    }
};

exports.getUserByUsername = (req, res) => {
    const { username } = req.query;
    // const { username } = req.body;

    Users.getByUsername(username, (err, data) => {
        if (err) return res.status(500).send(err);
        if (!data) return res.status(404).send('User not found');
        res.status(200).send(data);
    });
};

exports.updateUser = (req, res) => {
    const { old_username, username, email} = req.body;

    const updatedUser = {
        username: username,
        email: email,
    };

    Users.update(old_username, updatedUser, (err, data) => {
        if (err) return res.status(500).send(err);
        res.status(200).send('User updated successfully');
    });
};

File: backend-go/controllers/userController.go
// controllers/userController.go
package controllers

import (
	"backend-go/models"
	"database/sql"
	"encoding/json"
	"net/http"

	"golang.org/x/crypto/bcrypt"
)

// Register creates a new user with bcrypt hashed password and salt
func Register(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var user models.User
		err := json.NewDecoder(r.Body).Decode(&user)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		// Generate bcrypt hashed password and salt
		err = user.GeneratePasswordHash()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		// Create user in database
		err = user.CreateUser(db)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusCreated)
		json.NewEncoder(w).Encode(user)
	}
}

// Login verifies user credentials
func Login(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var user models.User
		err := json.NewDecoder(r.Body).Decode(&user)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		// Find user by username
		storedUser, err := models.FindUserByUsername(db, user.Username)
		if err != nil {
			http.Error(w, "Invalid username or password", http.StatusUnauthorized)
			return
		}

		// Compare hashed passwords
		err = bcrypt.CompareHashAndPassword([]byte(storedUser.Password), []byte(user.Password))
		if err != nil {
			http.Error(w, "Invalid username or password", http.StatusUnauthorized)
			return
		}

		// Passwords match, proceed with login
		json.NewEncoder(w).Encode(storedUser)
	}
}

// GetUserByUsername fetches a user by username
func GetUserByUsername(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		username := r.URL.Query().Get("username")

		user, err := models.GetUserByUsername(db, username)
		if err != nil {
			http.Error(w, "User not found", http.StatusNotFound)
			return
		}

		json.NewEncoder(w).Encode(user)
	}
}

// UpdateUser updates a user's username and email
func UpdateUser(db *sql.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var updateData struct {
			OldUsername string `json:"old_username"`
			Username    string `json:"username"`
			Email       string `json:"email"`
		}

		err := json.NewDecoder(r.Body).Decode(&updateData)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}

		var user models.User
		user.Username = updateData.Username
		user.Email = updateData.Email

		err = user.UpdateUser(db, updateData.OldUsername)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		json.NewEncoder(w).Encode(user)
	}
}
